### PR TITLE
Allow image streams to request pullthrough usage

### LINKED ISSUE #12214: Allow image stream consumers to opt in to pullthrough
Imported images can't use pullthrough because image resolution prefers the remote name.  Now that we have blob mirroring it should be possible for an image stream author to control how images are referenced.

Flag on image stream tag probably, defaults to false.  ResolveImageID attempts to convert to pullthrough.

Also need to make pullthrough default to true on config load.

### PR DESCRIPTION
**Pull request by smarterclayton**
Pullthrough (with blob mirroring) can ensure that a cluster is isolated from failures that occur in external registries, and also allows image providers to abstract access control (permissions are enforced inside the integrated registry and can thus hide the upstream credentials from consumers).

This PR adds a new `referencePolicy` struct to each image stream tag that allows an image stream author request that build config and deployment config change triggers use pullthrough (by specifying the `Local` TagReferencePolicyType). At resolution time, if the integrated registry is enabled the build or deployment config will use the local value (local registry + image id) instead of the external reference.

Changes CLI and adds extended tests.

Fixes #12214

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by mfojtik** [comment]
@smarterclayton i guess this also deserves a trello card

File: pkg/dockerregistry/server/repositorymiddleware.go [hunk 1]
```diff
@@ -178,7 +178,7 @@ func newRepositoryWithClient(
 	if err != nil {
 		context.GetLogger(ctx).Error(err)
 	}
-	pullthrough, err := getBoolOption(PullthroughEnvVar, "pullthrough", false, options)
+	pullthrough, err := getBoolOption(PullthroughEnvVar, "pullthrough", true, options)
```
Comment (mfojtik) [inline pkg/dockerregistry/server/repositorymiddleware.go]: does this break backward compatibility of this setting? iow. pullthrough is now enabled by default.

**Comment by mfojtik** [comment]
[(5) Allow image streams to request pullthrough usage](https://trello.com/c/8pPgFuJH/852-5-allow-image-streams-to-request-pullthrough-usage)

File: pkg/dockerregistry/server/repositorymiddleware.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/dockerregistry/server/repositorymiddleware.go]: We reserve the right to enable new features in a backwards compatible way (generally).  Pullthrough is as secure as our current image pull mechanism, so the original reasons for disabling this are no longer required.

File: pkg/image/api/types.go [hunk 2]
```diff
@@ -235,6 +235,8 @@ type TagReference struct {
 	Generation *int64
 	// ImportPolicy is information that controls how images may be imported by the server.
 	ImportPolicy TagImportPolicy
+	// ReferencePolicy defines how other components should consume the image
+	ReferencePolicy TagReferencePolicy `json:"referencePolicy,omitempty"`
```
Comment (miminar) [inline pkg/image/api/types.go]: superfluous json tag

Comment (smarterclayton) [inline pkg/image/api/types.go]: Doh

**Comment by smarterclayton** [comment]
Rebased, comment addressed, [merge].

**Comment by smarterclayton** [comment]
Looks like flake - triggers should run the same number of times as before:

```
--- FAIL: TestTriggers_configChange (5.03s)
	deploy_trigger_test.go:666: Operation cannot be fulfilled on deploymentconfigs "config": the object has been modified; please apply your changes to the latest version and try again
	etcd.go:99: dumping etcd to "/tmp/openshift/test-integration//etcd-dump-runtime.call32.json"
FAIL
```

[merge]

**Pull request merged**